File: src/launcher/FabricSupport.java
package launcher;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Minimal Fabric loader support: fetches the Fabric profile JSON (which is
 * itself a valid "version json" that inherits from the vanilla version) and
 * downloads its extra libraries. The merged result behaves like an
 * installed version whose id is "<mcVersion>-fabric-<loaderVersion>".
 */
public class FabricSupport {

    private static final String META_BASE = "https://meta.fabricmc.net/v2/versions/loader/";

    public static List<String> listLoaderVersions(String mcVersion) throws IOException, InterruptedException {
        String body = Http.getString(META_BASE + mcVersion);
        List<Object> arr = Json.asArray(Json.parse(body));
        List<String> out = new ArrayList<>();
        for (Object o : arr) {
            Map<String, Object> entry = Json.asObject(o);
            Map<String, Object> loader = Json.asObject(entry.get("loader"));
            String version = Json.asString(loader.get("version"), null);
            if (version != null) out.add(version);
        }
        return out;
    }

    /**
     * Installs Fabric on top of an already-installed vanilla version.
     * Returns the synthetic version id to launch.
     */
    public static String install(Path mcDir, String mcVersion, String loaderVersion,
                                  MinecraftInstaller.StatusListener listener) throws Exception {
        String profileUrl = META_BASE + mcVersion + "/" + loaderVersion + "/profile/json";
        String profileText = Http.getString(profileUrl);
        Map<String, Object> profile = Json.parseObject(profileText);

        String fabricId = mcVersion + "-fabric-" + loaderVersion;
        Path versionDir = mcDir.resolve("versions").resolve(fabricId);
        Files.createDirectories(versionDir);

        // Fabric libraries use {name, url} maven coordinates rather than
        // Mojang's {downloads:{artifact:{url,path}}} shape - resolve them here.
        Path librariesDir = mcDir.resolve("libraries");
        List<Object> libraries = Json.asArray(profile.get("libraries"));
        int i = 0;
        for (Object lo : libraries) {
            i++;
            Map<String, Object> lib = Json.asObject(lo);
            String name = Json.asString(lib.get("name"), null);
            String repoUrl = Json.asString(lib.get("url"), "https://maven.fabricmc.net/");
            if (name == null) continue;
            String path = mavenCoordToPath(name);
            Path dest = librariesDir.resolve(path);
            if (!Files.exists(dest)) {
                if (listener != null) listener.onStatus("Fabric library " + i + "/" + libraries.size(), i, libraries.size());
                String url = repoUrl.endsWith("/") ? repoUrl + path : repoUrl + "/" + path;
                try {
                    Http.download(url, dest, null);
                } catch (IOException e) {
                    System.out.println("Warning: failed to download fabric lib " + name + ": " + e.getMessage());
                }
            }
        }

        // Persist a version json for this synthetic id so it shows as installed;
        // it inherits the vanilla jar/assets and points at Fabric's main class.
        profile.put("inheritsFrom", mcVersion);
        Files.writeString(versionDir.resolve(fabricId + ".json"), Json.write(profile));

        return fabricId;
    }

    private static String mavenCoordToPath(String coord) {
        return MavenUtil.coordToPath(coord);
    }
}

File: electron/lib/fabricSupport.ts
import fs from 'fs';
import path from 'path';
import { getString, download } from './http';
import { coordToPath } from './mavenUtil';
import type { StatusListener } from './minecraftInstaller';

const META_BASE = 'https://meta.fabricmc.net/v2/versions/loader/';

export async function listLoaderVersions(mcVersion: string): Promise<string[]> {
  const body = await getString(META_BASE + mcVersion);
  const arr = JSON.parse(body);
  const out: string[] = [];
  for (const entry of arr) {
    const version = entry?.loader?.version;
    if (version) out.push(version);
  }
  return out;
}

/**
 * Installs Fabric on top of an already-installed vanilla version.
 * Returns the synthetic version id to launch.
 */
export async function install(
  mcDir: string,
  mcVersion: string,
  loaderVersion: string,
  listener?: StatusListener
): Promise<string> {
  const profileUrl = `${META_BASE}${mcVersion}/${loaderVersion}/profile/json`;
  const profileText = await getString(profileUrl);
  const profile = JSON.parse(profileText);

  const fabricId = `${mcVersion}-fabric-${loaderVersion}`;
  const versionDir = path.join(mcDir, 'versions', fabricId);
  fs.mkdirSync(versionDir, { recursive: true });

  // Fabric libraries use {name, url} maven coordinates rather than
  // Mojang's {downloads:{artifact:{url,path}}} shape - resolve them here.
  const librariesDir = path.join(mcDir, 'libraries');
  const libraries: any[] = profile.libraries ?? [];
  let i = 0;
  for (const lib of libraries) {
    i++;
    const name = lib?.name;
    const repoUrl = lib?.url ?? 'https://maven.fabricmc.net/';
    if (!name) continue;
    const relPath = coordToPath(name);
    if (!relPath) continue;
    const dest = path.join(librariesDir, relPath);
    if (!fs.existsSync(dest)) {
      listener?.(`Fabric library ${i}/${libraries.length}`, i, libraries.length);
      const url = repoUrl.endsWith('/') ? repoUrl + relPath : `${repoUrl}/${relPath}`;
      try {
        await download(url, dest, null);
      } catch (e: any) {
        console.log(`Warning: failed to download fabric lib ${name}: ${e.message}`);
      }
    }
  }

  // Persist a version json for this synthetic id so it shows as installed;
  // it inherits the vanilla jar/assets and points at Fabric's main class.
  profile.inheritsFrom = mcVersion;
  fs.writeFileSync(path.join(versionDir, `${fabricId}.json`), JSON.stringify(profile));

  return fabricId;
}
